design-proposals/2022-11-29-metal-port.md

```diff
@@ -0,0 +1,447 @@
+
+What that means for planning is a lot of little stages and a gradual hike up the difficulty curve.  We start with the simpler stuff and work our way to the more complex and we emphasize getting something to look at sooner.
+
+That sounds a bit scary!  But don't despair.  Once the developers get the guardrails in place and have it basically working, new problems tend to be fairly rational.  You don't get the deeply strange crashes unless you're doing something weird.
+
+We're also going to deviate a bit from the format of the Renderer Modularization Design proposal.  In that case we broke each section into *Required Changes* and *Benefits*.  The first set of these changes are all the same:  You have to do it to implement Metal.  The benefit is it works in Metal.  There's no practical mixing and matching with OpenGL.  If you want Metal support, someone has to do the first 10 of these tasks.
+
+### First, Do Nothing in Metal
+
+At the end of the Renderer Modularization work we'll have an OpenGL ES module for rendering to that SDK.  That will consist of a group of files, probably in a sub-directory, that can be included in the toolkit with a compile flag.
+
+Figure: OpenGL and Metal as plugins
+
+We'll want to do something similar for Metal.  We suggest building out the full set of files for that module in the most minimal way possible.  That is, a rendering loop that sets up, tears down, and renders nothing to the screen.  For textures, render targets, drawables and builders, again just stubs that do the minimum required to keep the toolkit from crashing.
+
+### Rendering Loop, Simple Metal Drawable, Simple Shader
+
+Once the renderer_impl module exists for Metal, it's time to start fleshing it out.  This is where the rendering "loop" begins, though subsequent development will change it.
+
+Figure: Metal Renderer (general)
+
+After the Rendering Modularization rework we'll have Layers using Builders to emit Drawables.  Initially the Metal Drawables will be stubs, but we can get started with a carefully curated Metal Drawable that consists of one polygon.  It can be matched with a single Metal Shader that does nothing more than draw it.
+
+Getting the skeleton version of the Metal Renderer in place is the purpose of this work.  When it's going, the developers should see a single rectangle moving around the screen as the user pans.  Not terribly exciting, but an easy start and a good place to branch off work with multiple developers.
+
+### Textures
+
+The MapLibre Native toolkit already knows what a texture is, so this is the Metal Texture variant.  Metal textures don't deviate all that much from the OpenGL in concept, but the specifics very much do.
+
+Metal supports a whole host of texture formats that OpenGL doesn't (or does through extensions).  There are is the traditional RGBA, but there are also 32 bit float, or dual 16 bit float, or.... the list goes on and on.  We want to allow support for these without losing our minds representing them.
```

Comment:
I think I'll change this to "Metal supports a whole host of texture formats that MapLibre Native doesn't"

I wasn't going to address the non-2D textures.  If you'd like to make a case for those in this toolkit, I'd be interested in hearing it.

As for atlas generation, we have a dynamic texture atlas in WhirlyGlobe-Maply and it is pretty great.  I've resisted proposing one here because what already exists works well enough.

Rendering to surfaces of different types requires the shaders to have the right output, certainly in Metal.  But it's a very powerful tool, as you point out, for things like Heatmaps.